winlogbeat/sys/wineventlog/wineventlog_windows.go

```diff
@@ -403,35 +403,35 @@ func FormatEventString(
 	}
 
 	var bufferPtr *byte
-	if renderBuf != nil {
+	if len(renderBuf) > 0 {
 		bufferPtr = &renderBuf[0]
 	}
 
 	// EvtFormatMessage operates with WCHAR buffer, assuming the size of the buffer in characters.
 	// https://docs.microsoft.com/en-us/windows/win32/api/winevt/nf-winevt-evtformatmessage
-	var bufferNeeded uint32
+	var bufferUsed uint32
 	bufferSize := uint32(len(renderBuf) / 2)
 
-	err := _EvtFormatMessage(ph, eventHandle, 0, 0, nil, messageFlag, bufferSize, bufferPtr, &bufferNeeded)
-	if err != nil && err != windows.ERROR_INSUFFICIENT_BUFFER { //nolint:errorlint // This is an errno.
+	err := _EvtFormatMessage(ph, eventHandle, 0, 0, nil, messageFlag, bufferSize, bufferPtr, &bufferUsed)
+	if err != nil && !errors.Is(err, windows.ERROR_INSUFFICIENT_BUFFER) {
 		return fmt.Errorf("failed in EvtFormatMessage: %w", err)
 	} else if err == nil {
-		// Windows API returns a null terminated WCHAR C-style string in the buffer. bufferNeeded applies
-		// only when ERROR_INSUFFICIENT_BUFFER is returned. Luckily the UTF16ToUTF8Bytes/UTF16ToString
-		// functions stop at null termination. Note, as signaled in a comment at the end of this function,
-		// this behavior is bad for EvtFormatMessageKeyword as then the API returns a list of null terminated
-		// strings in the buffer (it's fine for now as we don't use this parameter value).
-		return common.UTF16ToUTF8Bytes(renderBuf, out)
+		// bufferUsed indicates the size used internally to render the message. When called with nil buffer
```

Comment:
Yes, that's why I opted to increase the scratch buffer significantly to avoid double rendering. The size of the scratch buffer is still insignificant as it's only one buffer per event channel. Anyway unused memory from the buffer will just get paged out at runtime.

No I didn't add any test for the truncation as with having the fallback of allocating memory from the pool this should always succeed.